### PR TITLE
[expo-modules-core] Make expo-fonts and vector-icons work on fabric

### LINKED ISSUE #21115: [@expo/vector-icons] - icons/icon font not loading correctly
### Summary

When using the package `@expo/vector-icons` with other `expo` packages in an ejected environment, not a single icon loads and instead a question mark is rendered (it is even coloured in the correct way).

I am using the latest version of the package in connection with the new `react-native` architecture 
on version 0.71.2. Fabric renderer and concurrency are additionally enabled.

The icons are displayed as following:
showcase

Android devices are not affected by this issue.

### What platform(s) does this occur on?

iOS

### Environment

[...]
npmPackages:
@expo/vector-icons: 13.0.0 => 13.0.0
@expo/webpack-config: 0.17.4 => 0.17.4
babel-preset-expo: 9.2.2 => 9.2.2
expo: 48.0.0-beta.0 => 48.0.0-beta.0
react: 18.2.0 => 18.2.0
react-dom: 18.2.0 => 18.2.0
react-native: 0.71.2 => 0.71.2
react-native-web: 0.18.12 => 0.18.12
npmGlobalPackages:
expo-cli: 6.0.8
Expo Workflow: bare

### Minimal reproducible example
```
import { MaterialCommunityIcons } from "@expo/vector-icons"

[...]

<MaterialCommunityIcons color="blue" name="close" size={24} />
```

### LINKED ISSUE #22358: expo-font not working in Fabric
### Minimal reproducible example

Try a basic expo-font example

### Summary

Font does not show up on new architecture, but does on old.

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.3.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.16.0 - ~/.nvm/versions/node/v18.16.0/bin/node
      Yarn: 1.22.15 - /opt/homebrew/bin/yarn
      npm: 9.5.1 - ~/.nvm/versions/node/v18.16.0/bin/npm
      Watchman: 2023.02.20.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.12.1 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.2, iOS 16.2, macOS 13.1, tvOS 16.1, watchOS 9.1
    IDEs:
      Android Studio: 2022.1 AI-221.6008.13.2211.9619390
      Xcode: 14.2/14C18 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^48.0.0 => 48.0.9 
      react: 18.2.0 => 18.2.0 
      react-native: 0.71.6 => 0.71.6 
    Expo Workflow: bare

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

Expo-fonts don't work on fabric.
Fixes https://github.com/expo/expo/issues/21115
Fixes https://github.com/expo/expo/issues/22358

# How

In packages/expo-modules-core/ios/Services/EXReactFontManager.m we swizzle a different function than currently to enable fabric to work. It was a pain to find 😅

# Test Plan

Wrote a tester app, before:
simulator_screenshot_16914C88-B2BC-47BC-9989-620E646CF2A5

after:
simulator_screenshot_F2463EDC-80CA-44F8-AE5F-AE90A7C30E49

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-modules-core/ios/Services/EXReactFontManager.m [hunk 1]
```diff
@@ -13,6 +13,25 @@
...
+  for (id<EXFontProcessorInterface> fontProcessor in currentFontProcessors) {
+    NSNumber *size = [NSNumber numberWithFloat:fontSize];
+    font = [fontProcessor updateFont:nil withFamily:name size:size weight:nil style:nil variant:nil scaleMultiplier:1];
+    if (font) {
+      return font;
+    }
+  }
+//  EXFont* font = EXFontManager.registry2[name];
+  return [self EXfontWithName:name size:fontSize];
+//
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Services/EXReactFontManager.m:30]:
```suggestion
  return [self EXfontWithName:name size:fontSize];
```

File: packages/expo-modules-core/ios/Services/EXReactFontManager.m [hunk 2]
```diff
@@ -94,6 +110,29 @@ + (void)initialize
...
+{
+  Class uiFont = [UIFont class];
+  SEL uiUpdate = @selector(fontWithName:size:);
+  SEL exUpdate = @selector(EXfontWithName:size:);
+
+  Method originalMethod = class_getClassMethod(uiFont, uiUpdate);
+  Method swizzledMethod = class_getClassMethod(uiFont, exUpdate);
+
+  method_exchangeImplementations(originalMethod, swizzledMethod);
+}
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Services/EXReactFontManager.m:124]: Maybe let's have just one method but with `#ifdef RN_FABRIC_ENABLED` to separate these implementations

File: packages/expo-modules-core/ios/Services/EXReactFontManager.m [hunk 3]
```diff
@@ -13,6 +13,25 @@
...
+  for (id<EXFontProcessorInterface> fontProcessor in currentFontProcessors) {
+    NSNumber *size = [NSNumber numberWithFloat:fontSize];
+    font = [fontProcessor updateFont:nil withFamily:name size:size weight:nil style:nil variant:nil scaleMultiplier:1];
+    if (font) {
+      return font;
+    }
+  }
+  //  EXFont* font = EXFontManager.registry2[name];
+  return [self EXfontWithName:name size:fontSize];
+  //
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Services/EXReactFontManager.m:30]:
```suggestion
  return [self EXfontWithName:name size:fontSize];
```

File: packages/expo-modules-core/ios/Services/EXReactFontManager.m [hunk 4]
```diff
@@ -30,7 +49,7 @@ + (UIFont *)EXUpdateFont:(UIFont *)uiFont
       return font;
     }
   }
-
+
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Services/EXReactFontManager.m:52]:
```suggestion

```

File: packages/expo-modules-core/ios/Services/EXReactFontManager.m [hunk 5]
```diff
@@ -80,18 +99,39 @@ - (instancetype)init
...
+{
+  
+}
+
+// A utility function used for swizzling UIFont for new architecture
++ (void)registerFontsForUIFont
+{
+  
+}
+
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Services/EXReactFontManager.m:113]:
```suggestion
```

File: packages/expo-modules-core/ios/Services/EXReactFontManager.m [hunk 6]
```diff
@@ -13,6 +13,23 @@
...
++ (nullable UIFont *)EXfontWithName:(NSString *)name size:(CGFloat)fontSize
+{
+  UIFont *font;
+  for (id<EXFontProcessorInterface> fontProcessor in currentFontProcessors) {
+    NSNumber *size = [NSNumber numberWithFloat:fontSize];
+    font = [fontProcessor updateFont:nil withFamily:name size:size weight:nil style:nil variant:nil scaleMultiplier:1];
+    if (font) {
+      return font;
+    }
+  }
```
Comment (lukmccall) [inline packages/expo-modules-core/ios/Services/EXReactFontManager.m:27]:
```suggestion
  for (id<EXFontProcessorInterface> fontProcessor in currentFontProcessors) {
    NSNumber *size = [NSNumber numberWithFloat:fontSize];
    UIFont *font = [fontProcessor updateFont:nil withFamily:name size:size weight:nil style:nil variant:nil scaleMultiplier:1];
    if (font) {
      return font;
    }
  }
```
The `font` variable isn't used outside of the for loop.

**Review by lukmccall — APPROVED** [review]

**Pull request merged**